### PR TITLE
fix: Zip artifacts before uploading

### PR DESCRIPTION
**Pull request by pranavrc**
# Context

Currently we're uploading each artifact in a separate network call, so a build with a lot of artifacts creates a lot of network calls. The PR is to zip all the artifacts into one file and upload it so that aws lambda can take care of unzipping it.

# Objective

- [x] Check if `SD_ZIP_ARTIFACTS` is set
- [x] If it is, zip all the artifacts and uploading it to the store

# Related links

Issue - https://github.com/screwdriver-cd/screwdriver/issues/1196
Adding core/zip from habitat - https://github.com/screwdriver-cd/launcher/pull/198

### CODE REVIEW

File: commands.txt [hunk 1]
```diff
@@ -1,4 +1,11 @@
 if [ `uname` = 'Darwin' ]; then export SD_CURL_CMD_WRAPPER="eval"; else export SD_CURL_CMD_WRAPPER="$(echo 'sd-step exec --pkg-version 7.54.1 core/curl')"; fi
 cd $SD_ARTIFACTS_DIR
 find . -type f -print > manifest.txt
-find . -type f -exec sh -c 'f=`echo $1 | sed "s/^\.\///"`; encoded=$($SD_CURL_CMD_WRAPPER "curl -s -w \"%{url_effective}\" --data-urlencode \"$f\" -G \"\" | sed -e \"s/^\/?//\""); $SD_CURL_CMD_WRAPPER "curl -H \"Authorization: Bearer $SD_TOKEN\" -H \"Content-Type: text/plain\" -X PUT $STORE_URL/v1/builds/$SD_BUILD_ID/$ARTIFACTS_DIR_SUFFIX/$encoded -T ./\"$f\""' -- {} \;
+
+if [ "$SD_ZIP_ARTIFACTS" = "yes" ]; then
```
Comment (tkyi) [inline commands.txt]: `"yes"`? should we just use `true`?

Comment (pranavrc) [reply commands.txt]: Yeah, either that or it can just be `if [ ! -z "$SD_ZIP_ARTIFACTS" ]; then`

**Comment by d2lam** [comment]
fix test?

File: commands.txt [hunk 2]
```diff
@@ -1,4 +1,13 @@
 if [ `uname` = 'Darwin' ]; then export SD_CURL_CMD_WRAPPER="eval"; else export SD_CURL_CMD_WRAPPER="$(echo 'sd-step exec --pkg-version 7.54.1 core/curl')"; fi
 cd $SD_ARTIFACTS_DIR
 find . -type f -print > manifest.txt
-find . -type f -exec sh -c 'f=`echo $1 | sed "s/^\.\///"`; encoded=$($SD_CURL_CMD_WRAPPER "curl -s -w \"%{url_effective}\" --data-urlencode \"$f\" -G \"\" | sed -e \"s/^\/?//\""); $SD_CURL_CMD_WRAPPER "curl -H \"Authorization: Bearer $SD_TOKEN\" -H \"Content-Type: text/plain\" -X PUT $STORE_URL/v1/builds/$SD_BUILD_ID/$ARTIFACTS_DIR_SUFFIX/$encoded -T ./\"$f\""' -- {} \;
+
+if [ "$SD_ZIP_ARTIFACTS" = "true" ]; then
+    ZIPFILE=${SD_BUILD_ID}_artifacts.zip
+    find . -type f -exec zip $ZIPFILE {} +
```
Comment (jithine) [inline commands.txt]: do we have `zip` in all images ?

Comment (pranavrc) [reply commands.txt]: Not sure. We can use `tar` instead

Comment (pranavrc) [reply commands.txt]: Uhh, looks like tar can't create zip files, and lambda only unpacks zip files. So we might need the `zip` command

Comment (jithine) [reply commands.txt]: you will need to run this in a hab step. See `SD_CURL_CMD_WRAPPER`

File: commands.txt [hunk 3]
```diff
@@ -1,4 +1,13 @@
 if [ `uname` = 'Darwin' ]; then export SD_CURL_CMD_WRAPPER="eval"; else export SD_CURL_CMD_WRAPPER="$(echo 'sd-step exec --pkg-version 7.54.1 core/curl')"; fi
 cd $SD_ARTIFACTS_DIR
 find . -type f -print > manifest.txt
-find . -type f -exec sh -c 'f=`echo $1 | sed "s/^\.\///"`; encoded=$($SD_CURL_CMD_WRAPPER "curl -s -w \"%{url_effective}\" --data-urlencode \"$f\" -G \"\" | sed -e \"s/^\/?//\""); $SD_CURL_CMD_WRAPPER "curl -H \"Authorization: Bearer $SD_TOKEN\" -H \"Content-Type: text/plain\" -X PUT $STORE_URL/v1/builds/$SD_BUILD_ID/$ARTIFACTS_DIR_SUFFIX/$encoded -T ./\"$f\""' -- {} \;
+
+if [ "$SD_ZIP_ARTIFACTS" = "true" ]; then
+    ZIPFILE=${SD_BUILD_ID}_artifacts.zip
```
Comment (jithine) [inline commands.txt]: this name should match the lambda one. `SD_ARTIFACT.zip`

File: commands.txt [hunk 4]
```diff
@@ -1,4 +1,13 @@
 if [ `uname` = 'Darwin' ]; then export SD_CURL_CMD_WRAPPER="eval"; else export SD_CURL_CMD_WRAPPER="$(echo 'sd-step exec --pkg-version 7.54.1 core/curl')"; fi
 cd $SD_ARTIFACTS_DIR
 find . -type f -print > manifest.txt
-find . -type f -exec sh -c 'f=`echo $1 | sed "s/^\.\///"`; encoded=$($SD_CURL_CMD_WRAPPER "curl -s -w \"%{url_effective}\" --data-urlencode \"$f\" -G \"\" | sed -e \"s/^\/?//\""); $SD_CURL_CMD_WRAPPER "curl -H \"Authorization: Bearer $SD_TOKEN\" -H \"Content-Type: text/plain\" -X PUT $STORE_URL/v1/builds/$SD_BUILD_ID/$ARTIFACTS_DIR_SUFFIX/$encoded -T ./\"$f\""' -- {} \;
+
+if [ "$SD_ZIP_ARTIFACTS" = "true" ]; then
```
Comment (winstonsimon) [inline commands.txt]: Lets ensure, we are not adding credentials (eg: aws/.credentials or .keys) files in to zip file. This credentials will be displayed in SDv4 dashboard and ppl can misuse it.

Comment (d2lam) [reply commands.txt]: We are only uploading what is inside `$SD_ARTIFACT_DIR` (https://github.com/screwdriver-cd/artifact-bookend/blob/6909c9f7b91ba6b0a13a977a4c90b55eef593955/commands.txt#L2). There shouldn't be any credentials there unless users intentionally move them there (which I think is user's fault if they do that)

**Review by d2lam — APPROVED** [review]

**Review by tkyi — APPROVED** [review]

File: test/data/commands.txt [hunk 5]
```diff
@@ -1 +1 @@
...
..._CMD_WRAPPER="eval" &&     export SD_ZIP_CMD_WRAPPER="" && else &&     export SD_CURL_CMD_WRAPPER="$(echo 'sd-step exec --pkg-version 7.54.1 core/curl')" &&     export SD_ZIP_CMD_WRAPPER="$(echo 'sd-step exec core/zip')" && fi &&  && cd $SD_ARTIFACTS_DIR && find . -type f -print > manifest.txt &&  && if [ "$SD_ZIP_ARTIFACTS" = "true" ]; then &&     ZIPFILE=SD_ARTIFACT.zip &&     find . -type f -exec $SD_ZIP_CMD_WRAPPER zip $ZIPFILE {} + &&     f=`echo $ZIPFILE | sed "s/^\.\///"` &&     encoded=$($SD_CURL_CMD_WRAPPER "curl -s -w \"%{url_effective}\" --data-urlencode \"$f\" -G \"\" | sed -e \"s/^\/?//\"") &&     $SD_CURL_CMD_WRAPPER "curl -H \"Authorization: Bearer $SD_TOKEN\" -H \"Content-Type: text/plain\" -X PUT https://store.screwdriver.cd/v1/builds/$SD_BUILD_ID/ARTIFACTS/$encoded -T ./\"$f\"" && else &&     find . -type f -exec sh -c 'f=`echo $1 | sed "s/^\.\///"`; encoded=$($SD_CURL_CMD_WRAPPER "curl -s -w \"%{url_effective}\" --data-urlencode \"$f\" -G \"\" | sed -e \"s/^\/?//\""); $SD_CURL_CMD_WRAPPER "curl -H \"Authorization: Bearer $SD_TOKEN\" -H \"Content-Type: text/plain\" -X PUT $STORE_URL/v1/builds/$SD_BUILD_ID/$ARTIFACTS_DIR_SUFFIX/$encoded -T ./\"$f\""' -- {} \; && fi
```
Comment (minzcmu) [inline test/data/commands.txt]: Does the script work? It's doing `then &&  export xxx`.....

File: commands.txt [hunk 6]
```diff
@@ -1,4 +1,20 @@
-if [ `uname` = 'Darwin' ]; then export SD_CURL_CMD_WRAPPER="eval"; else export SD_CURL_CMD_WRAPPER="$(echo 'sd-step exec --pkg-version 7.54.1 core/curl')"; fi
+if [ `uname` = 'Darwin' ]; then
+    export SD_CURL_CMD_WRAPPER="eval"
+    export SD_ZIP_CMD_WRAPPER=""
+else
```
Comment (minzcmu) [inline commands.txt]: This probably won't work, inside our code, we concat each newline with `&&`.
https://github.com/screwdriver-cd/artifact-bookend/blob/master/index.js#L42

**Review by minzcmu — CHANGES_REQUESTED** [review]

File: commands.txt [hunk 7]
```diff
@@ -1,4 +1,20 @@
-if [ `uname` = 'Darwin' ]; then export SD_CURL_CMD_WRAPPER="eval"; else export SD_CURL_CMD_WRAPPER="$(echo 'sd-step exec --pkg-version 7.54.1 core/curl')"; fi
+if [ `uname` = 'Darwin' ]; then
+    export SD_CURL_CMD_WRAPPER="eval"
+    export SD_ZIP_CMD_WRAPPER=""
```
Comment (jithine) [inline commands.txt]: we should set $SD_ZIP_ARTIFACTS to "no" in this case if we can't detect `zip` binary installed in the `Darwin` case

File: test/data/commands.txt (hunk 5 above)
Comment (pranavrc) [reply test/data/commands.txt]: fixed

File: commands.txt (hunk 6 above)
Comment (pranavrc) [reply commands.txt]: fixed

File: commands.txt (hunk 7 above)
Comment (pranavrc) [reply commands.txt]: Checked mac slaves, zip is available https://screwdriver.corp.yahoo.com/projects/216678 https://jenkins.mobile.screwdriver.corp.yahoo.com/jenkins/job/216678-v3-component/BUILD_CONTAINER=xcode9,label=IOS/4/console#console-section-0

**Review by minzcmu — APPROVED** [review]

**Review by jithine — APPROVED** [review]

**Pull request merged**